scipy/ndimage/_multimethods.py

```diff
@@ -0,0 +1,1065 @@
+import functools
+import numpy as np
+from scipy._lib.uarray import Dispatchable, all_of_type, create_multimethod
+from scipy.ndimage import _api
+
+
+__all__ = ['affine_transform', 'binary_closing', 'binary_dilation',
+           'binary_erosion', 'binary_fill_holes', 'binary_hit_or_miss',
+           'binary_opening', 'binary_propagation', 'black_tophat',
+           'center_of_mass', 'convolve', 'convolve1d', 'correlate',
+           'correlate1d', 'distance_transform_bf', 'distance_transform_cdt',
+           'distance_transform_edt', 'extrema', 'find_objects',
+           'fourier_ellipsoid', 'fourier_gaussian', 'fourier_shift',
+           'fourier_uniform', 'gaussian_filter', 'gaussian_filter1d',
+           'gaussian_gradient_magnitude', 'gaussian_laplace',
+           'generate_binary_structure', 'generic_filter', 'generic_filter1d',
+           'generic_gradient_magnitude', 'generic_laplace',
+           'geometric_transform', 'grey_closing', 'grey_dilation',
+           'grey_erosion', 'grey_opening', 'histogram', 'iterate_structure',
+           'label', 'labeled_comprehension', 'laplace', 'map_coordinates',
+           'maximum', 'maximum_filter', 'maximum_filter1d', 'maximum_position',
+           'mean', 'median', 'median_filter', 'minimum', 'minimum_filter',
+           'minimum_filter1d', 'minimum_position', 'morphological_gradient',
+           'morphological_laplace', 'percentile_filter', 'prewitt',
+           'rank_filter', 'rotate', 'shift', 'sobel', 'spline_filter',
+           'spline_filter1d', 'standard_deviation', 'sum', 'sum_labels',
+           'uniform_filter', 'uniform_filter1d', 'variance', 'watershed_ift',
+           'white_tophat', 'zoom']
+
+
+_mark_non_coercible = functools.partial(Dispatchable, dispatch_type=np.ndarray,
```

Comment:
This `_mark_non_coercible` is used for the `output` argument, but it cannot have a dispatch type of `np.ndarray` since we need to also support dtypes or strings that can represent dtypes.

e.g. output='f', output='float32', output=np.float32 are all acceptable in addition to arrays

I am not sure the best way to handle this. One option is just to define a class like

```Python
class ndimage_output:
    pass
```

and then set `dispatch_type=ndimage_output` and the `__ua_convert__` function could handle this using something like:

```Python
    if dispatch_type is ndimage_output:
        if isinstance(value, np.ndarray):
            return np.asarray(value)
        else:
            try:
                return np.dtype(str(value))
            except TypeError:
                return np.dtype(value)
```

Perhaps, @peterbell10 knows of a better way to handle this, though?

Reply:
Done!